internal/codespaces/connection/port_forwarder.go

```diff
@@ -0,0 +1,243 @@
+		err = fwd.connection.tunnelClient.WaitForForwardedPort(ctx, remotePort)
+		if err != nil {
+			done <- fmt.Errorf("wait for forwarded port failed: %v", err)
+			return
+		}
+
+		// Connect to the forwarded port via a local TCP port
+		err = fwd.connection.tunnelClient.ConnectToForwardedPort(ctx, listen, remotePort)
+		if err != nil {
+			done <- fmt.Errorf("connect to forwarded port failed: %v", err)
+			return
+		}
+	}()
+	for {
+		select {
+		case err := <-done:
+			if err != nil {
+				return fmt.Errorf("error connecting to tunnel: %w", err)
+			}
+			return nil
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// ListPorts fetches the list of ports that are currently forwarded.
+func (fwd *PortForwarder) ListPorts(ctx context.Context) (ports []*tunnels.TunnelPort, err error) {
+	ports, err = fwd.connection.tunnelManager.ListTunnelPorts(ctx, fwd.connection.tunnel, fwd.connection.options)
+	if err != nil {
+		return nil, fmt.Errorf("error listing ports: %w", err)
+	}
+
+	return ports, nil
+}
+
+// UpdatePortVisibility changes the visibility (private, org, public) of the specified port.
+func (fwd *PortForwarder) UpdatePortVisibility(ctx context.Context, remotePort int, visibility string) error {
+	tunnelPort, err := fwd.connection.tunnelManager.GetTunnelPort(ctx, fwd.connection.tunnel, remotePort, fwd.connection.options)
+	if err != nil {
+		return fmt.Errorf("error getting tunnel port: %w", err)
+	}
+
+	// If the port visibility isn't changing, don't do anything
+	if AccessControlEntriesToVisibility(tunnelPort.AccessControl.Entries) == visibility {
+		return nil
+	}
+
+	// Delete the existing tunnel port to update
+	err = fwd.connection.tunnelManager.DeleteTunnelPort(ctx, fwd.connection.tunnel, uint16(remotePort), fwd.connection.options)
+	if err != nil {
+		return fmt.Errorf("error deleting tunnel port: %w", err)
+	}
+
+	done := make(chan error)
+	go func() {
+		// Connect to the tunnel
+		err = fwd.connection.tunnelClient.Connect(ctx, "")
+		if err != nil {
+			done <- fmt.Errorf("connect failed: %v", err)
+			return
+		}
+
+		// Inform the host that we've deleted the port
+		err = fwd.connection.tunnelClient.RefreshPorts(ctx)
+		if err != nil {
+			done <- fmt.Errorf("refresh ports failed: %v", err)
+			return
+		}
+
+		done <- nil
+	}()
+
+	// Wait for the done channel to be closed
+	for {
+		select {
+		case err := <-done:
+			if err != nil {
+				return fmt.Errorf("error connecting to tunnel: %w", err)
+			}
+
+			// Re-forward the port with the updated visibility
+			err = fwd.ForwardPort(ctx, uint16(remotePort), nil, false, false, false, visibility)
+			if err != nil {
+				return fmt.Errorf("error forwarding port: %w", err)
+			}
+
+			return nil
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// AccessControlEntriesToVisibility converts the access control entries used by Dev Tunnels to a friendly visibility value.
+func AccessControlEntriesToVisibility(accessControlEntries []tunnels.TunnelAccessControlEntry) string {
```

Comment:
This feels like it should live in `pkg/cmd/codespace/ports.go` unless you foresee it being needed for other commands.

Reply:
It's being used within this file and is also specific to this implementation of the port forwarder so I think it's better to leave where it is.